### PR TITLE
Add events emitting

### LINKED ISSUE #143: Add support for events of VirtualMachineImport objects
It would be very useful for debugging (and visibility in the UI) to add events support. These should at least include important changes in Conditions and errors.

### PR DESCRIPTION
**Pull request by machacekondra**
This PR adds event emitting for the VM import flow.
Fixes: https://github.com/kubevirt/vm-import-operator/issues/143

### CODE REVIEW

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go [hunk 1]
```diff
@@ -255,6 +269,9 @@ func (r *ReconcileVirtualMachineImport) importDisks(provider provider.Provider,
 					if err := r.afterSuccess(vmName, vmiName, provider); err != nil {
 						return err
 					}
+
+					// Emit event vm is successfully imported
+					r.recorder.Event(instance, corev1.EventTypeNormal, EventImportSucceeded, "Virtual Machine import successful")
```
Comment (pkliczewski) [inline pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: it would be good to add vm name so the user would know which vm was imported.
I would rephrase it to "Virtual Machine %s was imported successfully"

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go [hunk 2]
```diff
@@ -314,8 +331,12 @@ func (r *ReconcileVirtualMachineImport) createVM(provider provider.Provider, ins
 	// Create kubevirt VM from source VM:
 	reqLogger.Info("Creating a new VM", "VM.Namespace", vmSpec.Namespace, "VM.Name", vmSpec.Name)
 	if err = r.client.Create(context.TODO(), vmSpec); err != nil && !errors.IsAlreadyExists(err) {
+		message := fmt.Sprintf("Error while creating virtual machine: %s", err)
```
Comment (pkliczewski) [inline pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: Can we add vm name here as well?

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go [hunk 3]
```diff
@@ -362,10 +383,15 @@ func (r *ReconcileVirtualMachineImport) startVM(provider provider.Provider, inst
 				return err
 			}
 			if err = r.updateToRunning(vmName); err != nil {
+				// Emit event vm failed to start:
+				r.recorder.Eventf(instance, corev1.EventTypeWarning, EventVMStartFailed, "Virtual Machine failed to start: %s", err)
```
Comment (pkliczewski) [inline pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: Which vm failed to start?

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go [hunk 4]
```diff
@@ -445,6 +471,7 @@ func (r *ReconcileVirtualMachineImport) createDataVolumes(provider provider.Prov
 	// Fetch VM:
 	vmDef := &kubevirtv1.VirtualMachine{}
 	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: vmName.Namespace, Name: vmName.Name}, vmDef)
+
```
Comment (pkliczewski) [inline pkg/controller/virtualmachineimport/virtualmachineimport_controller.go:474]: I think that this blank line is not needed.

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go [hunk 5]
```diff
@@ -457,8 +484,12 @@ func (r *ReconcileVirtualMachineImport) createDataVolumes(provider provider.Prov
 
 		err = r.client.Create(context.TODO(), &dv)
 		if err != nil {
+			message := fmt.Sprintf("Data volume creation failed: %s", err)
```
Comment (pkliczewski) [inline pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: Would be good to add a bit of context about which datavolume for which vm.

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go [hunk 6]
```diff
@@ -821,6 +853,10 @@ func (r *ReconcileVirtualMachineImport) validate(instance *v2vv1alpha1.VirtualMa
 		}
 		if valid, message := shouldFailWith(conditions); !valid {
 			logger.Info("Import blocked. " + message)
+
+			// Emit event vm import blocked:
+			r.recorder.Eventf(instance, corev1.EventTypeNormal, EventImportBlocked, "Virtual Machine import blocked: %s", message)
```
Comment (pkliczewski) [inline pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: please add vm name

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go (hunk 1 above)
Comment (masayag) [reply pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: pls use Namespace/Name for VM

File: pkg/controller/virtualmachineimport/virtualmachineimport_controller.go (hunk 6 above)
Comment (jakub-dzon) [reply pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: If the VM is blocked, that event will be emitted on every reconciliation until the cause for the block is fixed; I think we might need some throttling around recording of that event.

Comment (machacekondra) [reply pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: Well, we can see following in the events of vm import:
`
  Normal  ImportBlocked  1s (x23 over 79s)  virtualmachineimport-controller  Virtual Machine import blocked: VM uses unsupported bios type: i440fx_sea_bios
`

and following in get events:
`
50s         Normal    ImportBlocked             virtualmachineimport/myvm                                    Virtual Machine import blocked: VM uses unsupported bios type: i440fx_sea_bios
`

I think it's fine, it's not flooded, you just see counter. Or I can check whether this event was already generated for this VM import and don't generate it if it was.

Comment (jakub-dzon) [reply pkg/controller/virtualmachineimport/virtualmachineimport_controller.go]: It does not look that bad. I wonder if in case of mass import, when there is some issue with, for example, resource mapping for tens or hundreds of VMs the event rate would be problematic. My gut tells me that for a good event processing service thousands of events per second should not be a problem.

**Review by pkliczewski — APPROVED** [review]

**Review by jakub-dzon — APPROVED** [review]

**Review by masayag — APPROVED** [review]
